### ISSUE TITLE
Unexpected variable assignment

### ISSUE THREAD

**Issue by zyro** [yuin/gopher-lua#216]
- [x] GopherLua is a Lua5.1 implementation. You should be familiar with Lua programming language. Have you read [Lua 5.1 reference manual](http://www.lua.org/manual/5.1/) carefully?
- [x] GopherLua is a Lua5.1 implementation.  In Lua, to keep it simple, **it is more important to remove functionalities rather than to add functionalities unlike other languages** . If you are going to introduce some new cool functionalities into the GopherLua code base and the functionalities can be implemented by existing APIs, It should be implemented as a library. 

Please answer the following before submitting your issue:

1. What version of GopherLua are you using?
`db9ae37725ec65ae634769088627d2d118faf06a`
2. What version of Go are you using?
`go version go1.11.4 darwin/amd64`
3. What operating system and processor architecture are you using?
`Mac OS X 10.13.6 17G4015`
4. What did you do?
Ran the script below, no other `.lua` files present.
```
local function bar()
  return "bar"
end

local function test(foo)
  local should_not_change
  print("type(should_not_change) is " .. type(should_not_change) .. " and value is " .. tostring(should_not_change))
  foo = (foo or bar())
  print("type(should_not_change) is " .. type(should_not_change) .. " and value is " .. tostring(should_not_change))
end

test(nil)
```
5. What did you expect to see?
The script should print:
```
type(should_not_change) is nil and value is nil
type(should_not_change) is nil and value is nil
```
6. What did you see instead?
But instead prints:
```
type(should_not_change) is nil and value is nil
type(should_not_change) is string and value is bar
```

Looks like the `foo = (foo or bar())` line causes an unexpected variable assignment. Neither `foo = bar()` nor `foo = (foo or "bar")` seem to have that effect. It seems it must be an `or` that triggers a function call *and* assigns to a variable from the function arguments.

**Comment by yuin** [comment]
@zyro 
Thanks for reporting this.

This bug had been caused by the register allocation strategy that aggressively saving registers. I've fixed this bug in the latest commit. Could you confirm if this bug is fixed?

**Comment by zyro** [comment]
@yuin I've tested and confirmed the issue appears to be fixed by 1e6e6e1918e02ddf02e667e88e8aa756942448c5. Thanks for the quick response!
